### ISSUE TITLE
Misleading error message for using self in a macro

### ISSUE THREAD

**Issue by whitequark** [rust-lang/rust#39057]
To reproduce:
```rust
macro_rules! x {
    () => (self)
}

struct X;
impl X {
    fn y(&self) {
        x!()
    }
}
```
```
error[E0424]: `self` is not available in a static method
 --> <anon>:2:12
  |
2 |     () => (self)
  |            ^^^^ not available in static method
...
8 |         x!()
  |         ---- in this macro invocation
  |
  = note: maybe a `self` argument is missing?
```

**Comment by estebank** [comment]
[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c59f4d0e6ae3d22408d8f0f961689b1c):

```
error[E0424]: expected value, found module `self`
 --> src/lib.rs:2:12
  |
2 |     () => (self)
  |            ^^^^ `self` value is a keyword only available in methods with `self` parameter
...
8 |         x!()
  |         ---- in this macro invocation
```

**Comment by Aaron1011** [comment]
Current output:

```
error[E0424]: expected value, found module `self`
 --> src/lib.rs:2:12
  |
2 |       () => (self)
  |              ^^^^ `self` value is a keyword only available in methods with a `self` parameter
...
7 | /     fn y(&self) {
8 | |         x!()
  | |         ---- in this macro invocation
9 | |     }
  | |_____- this function has a `self` parameter, but a macro invocation can only access identifiers it receives from parameters
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I think this issue can be closed
